README.md
![JestMocks logo](https://i.imgur.com/aKhUcDx.png)
# 🎉 JestMocks
Collection of jest mocks to help you develop your projects faster.

## 📁 Submodules

**jest-googleapi-mock**
Use this module with  [Jest](https://facebook.github.io/jest/)  to run web tests that rely on  `@google/map`  package where you want to mock the API functions.

This module has no runtime dependencies so your project won't pull in additional module dependencies by using this.

**jest-ngredux-mock**
Use this module with [Jest](https://facebook.github.io/jest/) to run web tests that rely on `@angular-redux` package.

## 💾 How to install locally
Download the zipped files directly from the repository.

**By command line.** 
Open Git Bash and paste: ` git clone https://github.com/iGitScor/jest-mocks.git`

## 🧾 Licence
iGitScor/jest-mocks is licensed under the [MIT License.](https://github.com/iGitScor/jest-mocks/blob/master/LICENSE)


jest-googleapi-mock/CHANGELOG.md
# Change log
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com) and this project follows the [Semantic Versioning](http://semver.org) code.

## 0.2.1 - 2018-03-15
### Added
- Create invalid key mock

## 0.2.0 - 2018-03-15
### Added
- Create multiple mock responses

## 0.1.1 - 2018-02-28
### Changed
- Fix jest mock


## 0.1.0 - 2018-02-27
### Added
- Create geocode mock


jest-googleapi-mock/__tests__/geocode.test.js
jest.mock('@google/maps', () => require('../src/setup'));
const googleMapsClient = require('@google/maps').createClient({});

describe('@google/map', () => {
  beforeEach(() => {
    jest.clearAllMocks();
  });

  test('googleMapsClient.geocode', () => {
    return new Promise((resolve, reject) => {
      googleMapsClient.geocode({
        address: 'Test'
      }, (err, response) => {
        if (!err) {
          resolve(response.json.results);
        } else {
          reject(err);
        }
      });
    }).then(address => {
      expect(address[0].formatted_address).toEqual(
        '827 Chemin du Mas de Vignolles, 30900 Nîmes, France'
      );
    });
  });

  test('googleMapsClient.geocode', () => {
    return new Promise((resolve, reject) => {
      googleMapsClient.geocode({
        place_id: '42'
      }, (err, response) => {
        if (!err) {
          resolve(response.json.results);
        } else {
          reject(err);
        }
      });
    }).then(address => {
      expect(address[0].formatted_address).toEqual(
        '827 Chemin du Mas de Vignolles, 30900 Nîmes, France'
      );
    });
  });

  test('googleMapsClient.geocode empty', () => {
    return new Promise((resolve, reject) => {
      googleMapsClient.geocode({
        address: '🚫::-'
      }, (err, response) => {
        if (!err) {
          resolve(response.json.results);
        } else {
          reject(err);
        }
      });
    }).then(address => {
      expect(address).toEqual([]);
    });
  });

  test('googleMapsClient.geocode invalid request', () => {
    return new Promise((resolve, reject) => {
      googleMapsClient.geocode({}, (err, response) => {
        if (!err) {
          resolve(response.json.results);
        } else {
          reject(err);
        }
      });
    }).catch(error => {
      expect(error).toBeDefined();
      expect(error.status).toEqual(400);
    });
  });

  test('googleMapsClient invalid key', () => {
    const invalidClient = require('@google/maps').createClient({ key: '🚫' });
    expect(
      new Promise((resolve, reject) => {
        invalidClient.geocode({}, (err, response) => {
          console.log(err);
          if (!err) {
            resolve(response.json.results);
          } else {
            reject(err);
          }
        });
      }).catch(error => [error])
    ).resolves.toEqual(['Missing either a valid API key, or a client ID and secret']);
  })
});


jest-googleapi-mock/src/@google/map/geocode.js
import { INVALID_REQUEST } from './errors/invalid-request';
import { EMPTY } from './success/empty';
import { RESULT } from './success/result';

export class Geocode {
  constructor(jest, config) {
    Object.defineProperty(this, 'geocode', {
      enumerable: false,
      value: jest.fn((options, cb) => {
        if (config.key === '🚫') {
          throw 'Missing either a valid API key, or a client ID and secret';
        }
        if (
          options.address ||
          options.bounds ||
          options.components ||
          options.latlng ||
          options.place_id
        ) {
          let resultType = RESULT;
          if (options.address) {
            const parts = options.address.split('::');
            switch (parts[0]) {
              case '🚫':
                resultType = EMPTY;
                break;

              default:
                break;
            }
          }

          return cb(null, resultType);
        } else {
          return cb(INVALID_REQUEST);
        }
      })
    });
  }
}


jest-googleapi-mock/src/@google/map/success/empty.js
const EMPTY = {
  status: 200,
  headers: {
    'content-type': 'application/json; charset=UTF-8',
    date: 'Fri, 02 Mar 2018 15:51:07 GMT',
    expires: 'Sat, 03 Mar2018 15:51:07 GMT',
    'cache-control': 'public, max-age=86400',
    'access-control-allow-origin': '*',
    server: 'mafe',
    'x-xss-protection': '1; mode=block',
    'x-frame-options': 'SAMEORIGIN',
    'alt-svc':
      'hq=":443"; ma=2592000; quic=51303431; quic=51303339; quic=51303338; quic=51303337; quic=51303335,quic=":443"; ma=2592000; v="41,39,38,37,35"',
    'accept-ranges': 'none',
    vary: 'Accept-Language,Accept-Encoding',
    connection: 'close'
  },
  json: { results: [], status: 'ZERO_RESULTS' },
  requestUrl:
    'https://maps.googleapis.com/maps/api/geocode/json?address=🚫::-&key=GOOGLEKEY',
  query: { address: '🚫::-', key: 'GOOGLEKEY' }
};

export { EMPTY };


jest-googleapi-mock/src/@google/map/errors/invalid-request.js
const INVALID_REQUEST = {
  status: 400,
  headers: {
    'content-type': 'application/json; charset=UTF-8',
    date: 'Fri, 02 Mar 2018 15:43:10 GMT',
    pragma: 'no-cache',
    expires: 'Fri, 01 Jan 1990 00:00:00 GMT',
    'cache-control': 'no-cache, must-revalidate',
    'access-control-allow-origin': '*',
    server: 'mafe',
    'x-xss-protection': '1; mode=block',
    'x-frame-options': 'SAMEORIGIN',
    'alt-svc':
      'hq=":443"; ma=2592000; quic=51303431; quic=51303339; quic=51303338;quic=51303337; quic=51303335,quic=":443"; ma=2592000; v="41,39,38,37,35"',
    'accept-ranges': 'none',
    vary: 'Accept-Encoding',
    connection: 'close'
  },
  json: {
    error_message:
      "Invalid request. Missing the 'address', 'bounds', 'components', 'latlng' or 'place_id' parameter.",
    results: [],
    status: 'INVALID_REQUEST'
  },
  requestUrl: 'https://maps.googleapis.com/maps/api/geocode/json?key=GOOGLEKEY',
  query: { key: 'GOOGLEKEY' }
};

export { INVALID_REQUEST };


jest-googleapi-mock/src/@google/map/success/result.js
const RESULT = {
  status: 200,
  headers: {
    'content-type': 'application/json; charset=UTF-8',
    date: 'Fri, 02 Mar 2018 15:52:40 GMT',
    expires: 'Sat, 03 Mar2018 15:52:40 GMT',
    'cache-control': 'public, max-age=86400',
    'access-control-allow-origin': '*',
    server: 'mafe',
    'x-xss-protection': '1; mode=block',
    'x-frame-options': 'SAMEORIGIN',
    'alt-svc':
      'hq=":443"; ma=2592000; quic=51303431; quic=51303339; quic=51303338; quic=51303337; quic=51303335,quic=":443"; ma=2592000; v="41,39,38,37,35"',
    'accept-ranges': 'none',
    vary: 'Accept-Language,Accept-Encoding',
    connection: 'close'
  },
  json: {
    results: [
      {
        address_components: [
          { long_name: '827', short_name: '827', types: ['street_number'] },
          {
            long_name: 'Chemin du Mas de Vignolles',
            short_name: 'Chemin du Mas de Vignolles',
            types: ['route']
          },
          {
            long_name: 'Nîmes',
            short_name: 'Nîmes',
            types: ['locality', 'political']
          },
          {
            long_name: 'Gard',
            short_name: 'Gard',
            types: ['administrative_area_level_2', 'political']
          },
          {
            long_name: 'Occitanie',
            short_name: 'Occitanie',
            types: ['administrative_area_level_1', 'political']
          },
          {
            long_name: 'France',
            short_name: 'FR',
            types: ['country', 'political']
          },
          { long_name: '30900', short_name: '30900', types: ['postal_code'] }
        ],
        formatted_address:
          '827 Chemin du Mas de Vignolles, 30900 Nîmes, France',
        geometry: {
          location: { lat: 43.8158633, lng: 4.363384900000001 },
          location_type: 'RANGE_INTERPOLATED',
          viewport: {
            northeast: { lat: 43.8172122802915, lng: 4.364733880291503 },
            southwest: { lat: 43.8145143197085, lng: 4.362035919708498 }
          }
        },
        partial_match: true,
        place_id:
          'EjQ4MjcgQ2hlbWluIGR1IE1hcyBkZSBWaWdub2xsZXMsIDMwOTAwIE7Drm1lcywgRnJhbmNl',
        types: ['street_address']
      }
    ],
    status: 'OK'
  },
  requestUrl:
    'https://maps.googleapis.com/maps/api/geocode/json?address=827%20Chemin%20du%20Mas%20de%20Vignolles%2C%2030000%20N%C3%AEmes%2C%20France&key=GOOGLEKEY',
  query: {
    address: '827 Chemin du Mas de Vignolles, 30000 Nîmes, France',
    key: 'GOOGLEKEY'
  }
};


export { RESULT };
